nixos/modules/services/mail/roundcube.nix

```diff
@@ -8,6 +8,11 @@ let
   localDB = cfg.database.host == "localhost";
   user = cfg.database.username;
   phpWithPspell = pkgs.php81.withExtensions ({ enabled, all }: [ all.pspell ] ++ enabled);
+  nginxCommonHeaders = lib.optionalString config.services.nginx.virtualHosts.${cfg.hostName}.forceSSL ''
+    add_header Strict-Transport-Security 'max-age=63072000; includeSubDomains';
+  '' + lib.optionalString (config.services.nginx.virtualHosts.${cfg.hostName}.quic && config.services.nginx.virtualHosts.${cfg.hostName}.http3) ''
+    add_header Alt-Svc 'h3=":$server_port"; ma=604800';
```

Comment:
`add_header Alt-Svc 'h3=":$server_port"; ma=86400';` is already done for `hasSSL && vhost.quic && vhost.http3` in the nginx module itself.

Also, it feels kinda arbitrary to implement behavior of the nginx module just for that vhost. Is there a reason for it, i.e. a suggestion from roundcube itself?

Reply:
This behavior is planned to be removed, setting headers requires a custom approach to configuring nginx.location.
Vore information - https://github.com/NixOS/nixpkgs/pull/275484#issuecomment-1863558456

Reply:
Allow this variant?
```diff
diff --git a/nixos/modules/services/mail/roundcube.nix b/nixos/modules/services/mail/roundcube.nix
index c35ece8362f6..c4394697e274 100644
--- a/nixos/modules/services/mail/roundcube.nix
+++ b/nixos/modules/services/mail/roundcube.nix
@@ -8,6 +8,11 @@ let
   localDB = cfg.database.host == "localhost";
   user = cfg.database.username;
   phpWithPspell = pkgs.php81.withExtensions ({ enabled, all }: [ all.pspell ] ++ enabled);
+  nginxCommonHeaders = lib.optionalString config.services.nginx.virtualHosts.${cfg.hostName}.forceSSL && config.services.nginx.virtualHosts.${cfg.hostName}.addHeaderHSTS ''
+    add_header Strict-Transport-Security 'max-age=63072000;';
+  '' + lib.optionalString (config.services.nginx.virtualHosts.${cfg.hostName}.quic && config.services.nginx.virtualHosts.${cfg.hostName}.http3) ''
+    add_header Alt-Svc 'h3=":$server_port"; ma=604800';
+  '';
 in
 {
   options.services.roundcube = {
diff --git a/nixos/modules/services/web-servers/nginx/vhost-options.nix b/nixos/modules/services/web-servers/nginx/vhost-options.nix
index 64a95afab9f4..b03794805497 100644
--- a/nixos/modules/services/web-servers/nginx/vhost-options.nix
+++ b/nixos/modules/services/web-servers/nginx/vhost-options.nix
@@ -133,6 +133,18 @@ with lib;
       '';
     };

+    addHeaderHSTS = mkOption {
+      type = types.bool;
+      default = false;
+      description = lib.mdDoc ''
+        Add an HTTP Strict-Transport-Security response header that informs browsers
+        that the site should only be accessed using HTTPS, and that any future attempts
+        to access it using HTTP should automatically be converted to HTTPS.
+        This option requires additional support from NixOS services that use nginx
+        as a web-server.
+      '';
+    };
+
     addSSL = mkOption {
       type = types.bool;
       default = false;
```

Reply:
> Unless patch nginx and change the behavior for managing headers.

And how does this fix the problem? You'll run into the exact same issue when adding custom headers to the roundcube vhost.

How to set the headers for QUIC and STS is something that should either be sorted out in the nginx module or it should be made clear that configuration for that is _always_ the admin's job.

By adding this to a bunch of random modules, you'll get a super inconsistent experience because you'll never know if the current module supports that or if you have to do it yourself (and follow-up question is, can you add custom headers to locations or do you need to add them to a vhost to avoid headers being discarded).

Even worse, I'd expect the implementation for each module which exposes STS to be slightly different which means that you'll need to double-check what exactly will happen for each vhost on your instance.

On the other hand, there are plenty of resources that show these headers need to look like and if you don't have to watch out for the extras a module does, I'd argue you're even quicker in setting things up.

Please note that the reasoning applies to both http3 and sts.

Reply:
> And how does this fix the problem? You'll run into the exact same issue when adding custom headers to the roundcube vhost.

If set the header once, nginx will add it to all places and exclude duplicates.


I want to choose the optimal configuration option so that most people do not need to configure anything additionally.
It’s difficult to do this on all services, but at least you can cover those that I use at myself.
If the current implementation option is not suitable for someone, then you can add an option to disable nginx automatic configuration.

Reply:
> If set the header once, nginx will add it to all places and exclude duplicates.

Am I missing something? I thought the headers aren't set once, but for each location?

> I want to choose the optimal configuration option so that most people do not need to configure anything additionally.

And that's exactly the end-result from your suggestion. People have to do stuff on their own when adding options like `addHeaderHSTS`. The way how this is done _will_ be totally inconsistent across all modules that use (or could use) nginx as reverse-proxy. The result is that it's easier to write down the header on your own rather than having too look up how $module can be configured into adding HSTS.

> but at least you can cover those that I use at myself

Then I don't see why this should be done upstream.
The module system is very good add extending existing configurations and that sounds exactly like what you need.

Reply:
> Am I missing something? I thought the headers aren't set once, but for each location?

If the header is set in the 'server' section, then it is inherited to all `location` sections. But, if one of `location` sections has a different header, then all headers from `server` section are removed. Therefore, the necessary headers from `server` section must be re-add in `location` section.

> hen I don't see why this should be done upstream.
> The module system is very good add extending existing configurations and that sounds exactly like what you need.

In current version of roundcube service, I cannot somehow change nginx configuration

> And that's exactly the end-result from your suggestion. People have to do stuff on their own when adding options like `addHeaderHSTS`. The way how this is done _will_ be totally inconsistent across all modules that use (or could use) nginx as reverse-proxy. The result is that it's easier to write down the header on your own rather than having too look up how $module can be configured into adding HSTS.

Ok I'll delete the HSTS header, and should I delete Alt-Svc too?
With use this PR, I will be able to add headers myself via `locations."/".extraConfig and `locations."~* \\.php(/|$)".extraConfig.

Reply:
> If the header is set in the 'server' section, then it is inherited to all location sections. But, if one of location sections has a different header, then all headers from server section are removed. Therefore, the necessary headers from server section must be re-add in location section.

I'm aware of that problem. What I didn't understand was why you meant that you'd need to set the headers once only.

> In current version of roundcube service, I cannot somehow change nginx configuration

Huh, sounds odd.
Would you mind sharing your unsuccessful attempt?

> Ok I'll delete the HSTS header, and should I delete Alt-Svc too?

That's what I'd suggest, yes.

Reply:
> I'm aware of that problem. What I didn't understand was why you meant that you'd need to set the headers once only.

I pointed out variant if there was a patch that would change how nginx works with headers.

> Huh, sounds odd.
> Would you mind sharing your unsuccessful attempt?

If activate roundcube service, it automatically configures nginx, and this cannot be disabled.
In this PR I added a `configureNginx` option that allows you to disable automatic configuration.

Reply:
> That's what I'd suggest, yes.

Removed this headers.

Reply:
>  In this PR I added a configureNginx option that allows you to disable automatic configuration.

fine for me :+1: